private/buf/bufsync/syncer.go

```diff
@@ -94,508 +76,686 @@ func newSyncer(
+	// ModuleBranches must be synced prior to ModuleTags. This ensures that all
+	// tagged commits in ModuleTags have been synced.
+	for _, moduleBranch := range plan.ModuleBranchesToSync() {
+		if err := s.handler.SyncModuleBranch(ctx, moduleBranch); err != nil {
+			return fmt.Errorf(
+				"sync module %s:%s branch %q: %w",
+				moduleBranch.Directory(),
+				moduleBranch.TargetModuleIdentity(),
+				moduleBranch.BranchName(),
+				err,
+			)
 		}
 	}
-	if err := s.syncTags(ctx); err != nil {
-		return fmt.Errorf("sync tags: %w", err)
+	for _, moduleTags := range plan.ModuleTagsToSync() {
+		if err := s.handler.SyncModuleTags(ctx, moduleTags); err != nil {
+			return fmt.Errorf(
+				"sync module %s tags: %w",
+				moduleTags.TargetModuleIdentity(),
+				err,
+			)
+		}
 	}
 	return nil
 }
 
-func (s *syncer) prepareSync(ctx context.Context) error {
-	// (1) Prepare all tags locations.
-	if err := s.repo.ForEachTag(func(tag string, commitHash git.Hash) error {
-		s.commitsToTags[commitHash.Hex()] = append(s.commitsToTags[commitHash.Hex()], tag)
-		return nil
-	}); err != nil {
-		return fmt.Errorf("load tags: %w", err)
+// resolveSyncPoint resolves a sync point for a target module identity and protected branch.
+// If there is no sync point for the branch, this returns (nil, nil).
+// The sync point is validated to exist before it is returned.
+func (s *syncer) resolveSyncPoint(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branch string,
+) (git.Hash, error) {
+	syncPoint, err := s.handler.ResolveSyncPoint(ctx, moduleIdentity, branch)
+	if err != nil {
+		return nil, fmt.Errorf("resolve sync point for module %s: %w", moduleIdentity.IdentityString(), err)
 	}
-	// (2) Prepare branches to be synced.
-	allBranches := make(map[string]struct{})
-	if err := s.repo.ForEachBranch(func(branch string, _ git.Hash) error {
-		allBranches[branch] = struct{}{}
-		return nil
-	}, git.ForEachBranchWithRemote(s.gitRemoteName)); err != nil {
-		return fmt.Errorf("looping over repository branches: %w", err)
+	if syncPoint == nil {
+		// No sync point for that module in that branch.
+		return nil, nil
+	}
+	// Validate that the commit pointed to by the sync point exists in the git repo.
+	if _, err := s.repo.Objects().Commit(syncPoint); err != nil {
+		// The most likely culprit for an invalid sync point is a rebase, where the last known commit has
+		// been garbage collected. In this case, let's present a better error message.
+		//
+		// This is not trivial scenario if the branch that's been rebased is a long-lived branch (like
+		// main) whose artifacts are consumed by other branches, as we may fail to sync those commits if
+		// we continue.
+		//
+		// For now we simply error with a specific message if this happens.
+		if errors.Is(err, git.ErrObjectNotFound) {
+			return nil, fmt.Errorf(
+				"last synced git commit %q for default branch %q in module %q is not found in the git repo, did you rebase or reset your default branch?",
```

Comment:
I think this func is only invoked for protected branches? Nit: rename func `s/resolveSyncPoint/resolveSyncPointForProtectedBranch/`, and `s/default/protected/` in this err message.

Reply:
Done